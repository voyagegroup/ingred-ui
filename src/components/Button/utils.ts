import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
