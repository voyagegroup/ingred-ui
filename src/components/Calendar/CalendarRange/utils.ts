import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
