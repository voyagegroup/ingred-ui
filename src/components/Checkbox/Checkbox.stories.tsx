import React from "react";
import { Story } from "@storybook/react/types-6-0";
import Checkbox from ".";
import { CheckBoxProps } from "./Checkbox";
import { Flex } from "..";

export default {
  title: "Components/Inputs/Checkbox",
  component: Checkbox,
  argTypes: {
    onChange: { action: "changed" },
  },
};

export const Example: Story<CheckBoxProps> = (args) => (
  <Checkbox {...args}>Checkbox</Checkbox>
);

export const DesignSamples = () => (
  <Flex display="flex" flexDirection="column">
    <Checkbox checked={false}>Not checked</Checkbox>
    <Checkbox checked={true}>Checked</Checkbox>
    <Checkbox indeterminate={true} checked={true}>
      Checked(indeterminate)
    </Checkbox>
    <Checkbox disabled checked={false}>
      Not checked(disabled)
    </Checkbox>
    <Checkbox disabled checked={true}>
      Checked(disabled)
    </Checkbox>
  </Flex>
);
