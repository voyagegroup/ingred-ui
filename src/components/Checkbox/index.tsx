export { default } from "./Checkbox";
export type { CheckBoxProps } from "./Checkbox";
