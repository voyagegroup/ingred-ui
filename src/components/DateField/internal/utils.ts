export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
