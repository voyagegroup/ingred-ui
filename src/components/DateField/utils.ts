import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
