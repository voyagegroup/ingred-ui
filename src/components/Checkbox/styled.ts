import styled from "styled-components";
import { hexToRgba } from "../../utils/hexToRgba";

export const Checkbox = styled.input<{
  indeterminate: boolean;
  error: boolean;
}>`
  opacity: 0;
  -webkit-appearance: none;
  appearance: none;
  position: absolute;

  &:checked + span::before {
    /* flex: 1 0 auto; */
    /* background */
    background-color: ${({ theme }) => theme.palette.primary.main};
    border-color: ${({ theme }) => theme.palette.primary.dark};

    /* check icon */
    background-image: ${({ indeterminate }) =>
      indeterminate
        ? `url('data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2210%22%20height%3D%223%22%20viewBox%3D%220%200%2010%203%22%3E%3Cdefs%3E%3Cstyle%3E.a%7Bfill%3A%23fff%3B%7D%3C%2Fstyle%3E%3C%2Fdefs%3E%3Crect%20class%3D%22a%22%20width%3D%2210%22%20height%3D%223%22%20rx%3D%220.596%22%2F%3E%3C%2Fsvg%3E')`
        : `url('data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2212.001%22%20height%3D%228.997%22%20viewBox%3D%220%200%2012.001%208.997%22%3E%3Cdefs%3E%3Cstyle%3E.a%7Bfill%3A%23fff%3B%7D%3C%2Fstyle%3E%3C%2Fdefs%3E%3Cpath%20class%3D%22a%22%20d%3D%22M405.681%2C343.44l-5.154%2C5.12-1.332%2C1.323a.622.622%2C0%2C0%2C1-.877%2C0l-1.333-1.323h0l-2.943-2.924a.612.612%2C0%2C0%2C1%2C0-.871l1.332-1.324a.624.624%2C0%2C0%2C1%2C.877%2C0l2.506%2C2.49%2C4.714-4.685a.624.624%2C0%2C0%2C1%2C.877%2C0l1.333%2C1.323A.614.614%2C0%2C0%2C1%2C405.681%2C343.44Z%22%20transform%3D%22translate(-393.861%20-341.066)%22%2F%3E%3C%2Fsvg%3E')`};
    background-repeat: no-repeat;
    background-position: center center;
  }
`;

export const Container = styled.label<{ disabled: boolean }>`
  cursor: ${({ disabled }) => (disabled ? "not-allowed" : "pointer")};
`;

export const Span = styled.span<{
  hasChild: boolean;
  indeterminate: boolean;
  error: boolean;
}>`
  display: inline-flex;
  align-items: center;
  color: ${({ error }) =>
    error
      ? ({ theme }) => theme.palette.danger.main
      : ({ theme }) => theme.palette.black};
  &::before {
    flex-shrink: 0;
    display: block;
    content: "";
    width: 18px;
    height: 18px;
    border: 1px solid
      ${({ error, theme }) =>
        error ? theme.palette.danger.main : theme.palette.divider};
    border-radius: ${({ theme }) => theme.radius * 0.5}px;
    box-shadow: ${({ theme }) =>
      `0 -2px ${hexToRgba(theme.palette.black, 0.16)} inset, 0 2px ${hexToRgba(
        theme.palette.black,
        0.08,
      )}`};
    background-color: ${({ theme }) => theme.palette.background.default};
    margin-right: ${({ hasChild, theme }) =>
      hasChild ? `${theme.spacing / 2}px` : "auto"};
    transition: background-color 0.3s ease;
  }
  ${Checkbox}:disabled + & {
    color: ${({ theme }) => theme.palette.text.disabled};
  }
  ${Checkbox}:disabled + &::before {
    border: 1px solid ${({ theme }) => theme.palette.divider};
    background-color: ${({ theme }) => theme.palette.gray.light};
  }
  ${Checkbox}:disabled:checked + &::before {
    background-image: ${({ indeterminate }) =>
      indeterminate
        ? `url('data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2210%22%20height%3D%223%22%20viewBox%3D%220%200%2010%203%22%3E%3Cdefs%3E%3Cstyle%3E.a%7Bfill%3A%23b3bac1%3B%7D%3C%2Fstyle%3E%3C%2Fdefs%3E%3Crect%20class%3D%22a%22%20width%3D%2210%22%20height%3D%223%22%20rx%3D%220.596%22%2F%3E%3C%2Fsvg%3E')`
        : `url('data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20width%3D%2212.001%22%20height%3D%228.997%22%20viewBox%3D%220%200%2012.001%208.997%22%3E%3Cdefs%3E%3Cstyle%3E.a%7Bfill%3A%23b3bac1%3B%7D%3C%2Fstyle%3E%3C%2Fdefs%3E%3Cpath%20class%3D%22a%22%20d%3D%22M405.681%2C343.44l-5.154%2C5.12-1.332%2C1.323a.622.622%2C0%2C0%2C1-.877%2C0l-1.333-1.323h0l-2.943-2.924a.612.612%2C0%2C0%2C1%2C0-.871l1.332-1.324a.624.624%2C0%2C0%2C1%2C.877%2C0l2.506%2C2.49%2C4.714-4.685a.624.624%2C0%2C0%2C1%2C.877%2C0l1.333%2C1.323A.614.614%2C0%2C0%2C1%2C405.681%2C343.44Z%22%20transform%3D%22translate(-393.861%20-341.066)%22%2F%3E%3C%2Fsvg%3E')`};
    background-repeat: no-repeat;
    background-position: center center;
  }
`;
