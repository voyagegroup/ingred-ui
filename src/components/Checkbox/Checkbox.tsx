import * as React from "react";
import * as Styled from "./styled";

export type CheckBoxProps = React.ComponentPropsWithoutRef<"input"> & {
  indeterminate?: boolean;
  error?: boolean;
  inputRef?: React.Ref<HTMLInputElement>;
};

const Checkbox = React.forwardRef<HTMLLabelElement, CheckBoxProps>(
  (
    {
      children,
      indeterminate = false,
      error = false,
      disabled = false,
      inputRef,
      ...rest
    },
    ref,
  ) => {
    return (
      <Styled.Container ref={ref} disabled={disabled}>
        <Styled.Checkbox
          ref={inputRef}
          error={error}
          readOnly={true}
          type="checkbox"
          indeterminate={indeterminate}
          disabled={disabled}
          {...rest}
        />
        <Styled.Span
          error={error}
          indeterminate={indeterminate}
          hasChild={!!children}
        >
          {children}
        </Styled.Span>
      </Styled.Container>
    );
  },
);

export default Checkbox;
