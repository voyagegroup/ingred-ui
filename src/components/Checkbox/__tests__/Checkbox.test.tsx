import * as React from "react";
import "@testing-library/jest-dom";
import { cleanup } from "@testing-library/react";
import Checkbox from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Checkbox component testing", () => {
  afterEach(cleanup);

  test("Checkbox", () => {
    const { asFragment } = renderWithThemeProvider(<Checkbox />);
    expect(asFragment()).toMatchSnapshot();
  });
});
