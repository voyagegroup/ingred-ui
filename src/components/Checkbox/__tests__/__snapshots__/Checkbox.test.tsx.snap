// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox component testing Checkbox 1`] = `
<DocumentFragment>
  <label
    class="sc-gsDKAQ eDBaTH"
  >
    <input
      class="sc-bdvvtL lbDrac"
      readonly=""
      type="checkbox"
    />
    <span
      class="sc-dkPtRN jHrIXp"
    />
  </label>
</DocumentFragment>
`;
